### PR TITLE
Pkcs11 create/destroy objects

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Depends on https://github.com/OP-TEE/optee_os/pull/3835

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Sorry, there seems to be some checkpatch errors. I'm fixing them now.
Please wait with review until I've fixed them.

**Comment by jenswi-linaro** [comment]
Done with the checkpatch errors. There are three left, but I'm afraid we'll need to accept those.

File: libckteec/src/serializer.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include "serializer.h"
+
+CK_RV init_serial_object(struct serializer *obj)
+{
+	struct pkcs11_object_head head = { };
+
+	memset(obj, 0, sizeof(*obj));
+
+	/* Init head to all ones, will be set at finalize_serial_object */
+	memset(&head, 0xFF, sizeof(head));
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serializer.c:22]: Good question, I'll drop this.

File: libckteec/src/serializer.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+	head.attrs_size = obj->size - sizeof(head);
+	head.attrs_count = obj->item_count;
+	memcpy(obj->buffer, &head, sizeof(head));
+}
+
+void release_serial_object(struct serializer *obj)
+{
+	free(obj->buffer);
+	obj->buffer = NULL;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serializer.c:38]: Yes, `release_serial_object()` might get called twice on the same `struct serializer` in for instance `ck_create_object()`.

File: libckteec/src/serializer.c [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+CK_RV serialize_buffer(struct serializer *obj, void *data, size_t size)
+{
+	return serialize(&obj->buffer, &obj->size, data, size);
+}
+
+CK_RV serialize_32b(struct serializer *obj, uint32_t data)
+{
+	uint32_t data32 = data;
+
+	return serialize_buffer(obj, &data32, sizeof(uint32_t));
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serializer.c:75]: I'll fix

File: libckteec/include/pkcs11_ta.h [hunk 4]
```diff
@@ -249,6 +249,35 @@ enum pkcs11_ta_cmd {
...
+	 * [in]  memref[0] = [
+	 *              32bit session handle,
+	 *              (struct pkcs11_object_head)attribs + attributes data
+	 *	 ]
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [out] memref[2] = 32bit object handle
+	 *
+	 * This command relates to the PKCS#11 API function C_CreateObject().
+	 */
+	PKCS11_CMD_IMPORT_OBJECT = 15,
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/include/pkcs11_ta.h:267]: @etienne-lms, do you know?

File: libckteec/src/ck_debug.c [hunk 5]
```diff
@@ -457,3 +454,4 @@ const char *ckk2str(CK_KEY_TYPE id)
 		return "Unknown";
 	}
 }
+
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/ck_debug.c:457]: OK

File: libckteec/src/serialize_ck.c [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+ * Copyright (c) 2017-2020, Linaro Limited
+ */
+
+#include <pkcs11.h>
+#include <pkcs11_ta.h>
+#include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
+#include <ck_debug.h>
+#include <pkcs11_ta.h>
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serialize_ck.c:12]: I'll fix

File: libckteec/src/serialize_ck.c [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+#include "ck_helpers.h"
+#include "local_utils.h"
+#include "serializer.h"
+#include "serialize_ck.h"
+
+/*
+ * Generic way of serializing CK keys, certif, mechanism parameters, ...
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serialize_ck.c:20]: I'll fix

File: libckteec/src/serialize_ck.c [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+ * In cryptoki 2.40 parameters are almost all packaged as struture below:
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serialize_ck.c:21]: Thanks

File: libckteec/src/serialize_ck.c [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+{
+	CK_ATTRIBUTE_PTR cur_attr = attributes;
+	CK_ULONG n = count;
+	CK_RV rv = CKR_OK;
+
+	rv = init_serial_object(obj);
+	if (rv)
+		return rv;
+
+	for (; n; n--, cur_attr++) {
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serialize_ck.c:184]: Or even better:
```C
        for (n = 0; n < count; n++) {
                rv = serialize_ck_attribute(obj, attributes + n);
                if (rv)
                        break;
        }
```
I'll fix

File: libckteec/include/pkcs11_ta.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libckteec/include/pkcs11_ta.h:267]: That's an old choice at the time more command ID names did fully match the related cryptoki API function labels. I think we can rename this to `PKCS11_CMD_CREATE_OBJECT` as suggested by @jforissier.

[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/include/pkcs11_ta.h:267]: OK, I'll fix

File: libckteec/src/serializer.c [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ * serialize - append data in a serialized buffer
+ *
+ * Serialize data in provided buffer.
+ * Ensure 64byte alignment of appended data in the buffer.
+ */
+static CK_RV serialize(char **bstart, size_t *blen, void *data, size_t len)
+{
+	char *buf;
+	size_t nlen;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply libckteec/src/serializer.c:48]: OK, I'll fix

**Comment by jenswi-linaro** [comment]
Squashed and tag applied. Updated the link to the optee_os PR.

**Comment by etienne-lms** [comment]
Indeed the 4th last commit still miss review tags. some are big pieces.
- ta: pkcs11: helper for serial arguments with allocation
- ta: pkcs11: persistent object support
- ta: pkcs11: attribute helper functions
- ta: pkcs11: support command to import and destroy object

In last "ta: pkcs11: support command to import and destroy object" commit log:
s/PKCS11_CMD_IMPORT_OBJECT/PKCS11_CMD_CREATE_OBJECT/

**Pull request merged**